src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -321,6 +327,344 @@ public async Task TestRemoteWorkspaceCircularReferences()
             Assert.NotNull(solution);
         }
 
+        private static ImmutableArray<ImmutableArray<T>> Permute<T>(T[] values)
+        {
+            using var _ = ArrayBuilder<ImmutableArray<T>>.GetInstance(out var result);
+            DoPermute(0, values.Length - 1);
+            return result.ToImmutable();
+
+            void DoPermute(int start, int end)
+            {
+                if (start == end)
+                {
+                    // We have one of our possible n! solutions,
+                    // add it to the list.
+                    result.Add(values.ToImmutableArray());
+                }
+                else
+                {
+                    for (var i = start; i <= end; i++)
+                    {
+                        (values[start], values[i]) = (values[i], values[start]);
+                        DoPermute(start + 1, end);
+                        (values[start], values[i]) = (values[i], values[start]);
+                    }
+                }
+            }
+        }
+
+        private static async Task TestInProcAndRemoteWorkspace(
+            params ImmutableArray<(string hintName, SourceText text)>[] values)
+        {
+            await TestInProcAndRemoteWorkspace(syncWithRemoteServer: true, values);
+        }
+
+        private static async Task TestInProcAndRemoteWorkspace(
+            bool syncWithRemoteServer, params ImmutableArray<(string hintName, SourceText text)>[] values)
+        {
+            // Try every permutation of these values.
+            foreach (var permutation in Permute(values))
+            {
+                var gotException = false;
+                try
+                {
+                    await TestInProcAndRemoteWorkspaceWorker(syncWithRemoteServer, permutation);
+                }
+                catch (XunitException)
+                {
+                    gotException = true;
+                }
+
+                // If we're syncing to the remove server, we should get no exceptions, since the data should be matched
+                // between both.  If we're not syncing, we should see a failure since the two processes will disagree on
+                // the contents.
+                Assert.Equal(syncWithRemoteServer, !gotException);
```

Comment:
each test actually tests with the logic enabled/disabled to actually sync the contents between the hosts, versus running the generator locally in each process.  We guarantee that if we're not syncing, we get different results (some test assertion fails), but we are syncing that everything was ok.